### PR TITLE
Add partial distinct for global aggergation without marking

### PR DESCRIPTION
**Pull request by dain**
This addresses item 2 from issue #10305 

This is for non group-by aggregations only.  Basically, we have two way to execute agg(distinct x) today.  One is you redistribute the table on `x` and mark which rows are distinct for `x` on N machines, and after all aggregations have marked stuff distinct, we can do a partial on N machines.  Alternatively, we can simply move all rows to one machine, and only do the distinct and all work on one machine.  With the mark distinct strategy you can do huge distincts, but it is super expensive, because you have to redistribute the table for each distinct aggregation.  The single strategy removes the redistributions, but can be high latency because all of work is one on one machine in one thread, and you only get single machine's memory (not distribute memory) for the distinct set.

This change, reduces the latency of the single aggregation by doing a partial mark distinct on each node, so you get N machines doing the marking. And, you get partials for the non-distinct aggregations.  You still have the problem where the final distinct set does not fit in single machine's memory.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java [hunk 1]
```diff
@@ -82,6 +86,14 @@ public AggregationNode(
...
                 .forEach(outputSymbols::add);
         this.outputSymbols = outputSymbols.build();
+
+        ImmutableSet.Builder<Symbol> inputSymbols = ImmutableSet.builder();
+        inputSymbols.addAll(getGroupingKeys());
+        hashSymbol.ifPresent(inputSymbols::add);
+        aggregations.stream()
+                .map(Aggregation::getInputSymbols)
+                .forEach(inputSymbols::addAll);
+        this.inputSymbols = inputSymbols.build();
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: @dain Should we include `groupIdSymbol` into `inputSymbols`?

File: presto-spi/src/main/java/com/facebook/presto/spi/Page.java [hunk 2]
```diff
@@ -288,6 +288,11 @@ public Page prependColumn(Block column)
         return new Page(positionCount, result);
     }
 
+    public Page selectChannels(List<Integer> channels)
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/Page.java]: @dain This would be useful in multiple places. For example, `com.facebook.presto.operator.WindowOperator#rearrangePage` and  `com.facebook.presto.operator.StreamingAggregationOperator#extractColumns`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java (hunk 1 above)
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: I believe `getGroupingKeys()` above includes it

File: presto-spi/src/main/java/com/facebook/presto/spi/Page.java (hunk 2 above)
Comment (dain) [reply presto-spi/src/main/java/com/facebook/presto/spi/Page.java]: When I added it, I searched the code and couldn't find more places.  I'll add these when I update this PR.  If you know about other places, let me know.

**Comment by dain** [comment]
The approach has a problem with the `PushPartialAggregationThroughJoin` optimizer, which assumes that a partial global aggregation can be converted to a partial grouped aggregation, and the new code simply does not support this.

**Comment by mbasmanova** [comment]
@dain Thanks for explaining. Test failures do mention that rule:

```
Caused by: java.lang.IllegalArgumentException: com.facebook.presto.sql.planner.iterative.rule.PushPartialAggregationThroughJoin: transformed expression doesn't produce same outputs: [rowtype, count_44] vs [count_44]
```

**Comment by mbasmanova** [comment]
@dain Dain, I noticed that commits here appear out of order. Do you now how to fix that? @findepi Piotr had some magic git command, but I can't find it.

**Comment by dain** [comment]
@mbasmanova fixed... it is `git rebase --ignore-date`

**Comment by findepi** [comment]
i think `--ignore-date` is (was?) not always enough, at least I needed to add sleeps, so that timestamps are unique -- https://gist.github.com/findepi/7abf17c8a26d3b74e8bb9527caadfdbe

(Maybe it's no longer necessary though. https://github.com/prestodb/presto/pull/10840/commits shows in order, even though https://github.com/prestodb/presto/pull/10840#commits-pushed-100411b does not)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 3]
```diff
@@ -2413,17 +2419,22 @@ else if (target instanceof DeleteHandle) {
...
-        }
+            if (Arrays.equals(channels, range(0, inputLayout.size()).toArray())) {
+                // this is an identity mapping
+                return Function.identity();
+            }
 
-        return new PageChannelSelector(channels);
+            return new PageChannelSelector(channels);
+        }
+        catch (Exception e) {
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: Why is this needed?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java [hunk 4]
```diff
@@ -78,8 +80,13 @@ public AggregationNode(
         this.step = requireNonNull(step, "step is null");
         this.hashSymbol = requireNonNull(hashSymbol, "hashSymbol is null");
         this.groupIdSymbol = requireNonNull(groupIdSymbol, "groupIdSymbol is null");
+        this.rowTypeSymbol = requireNonNull(rowTypeSymbol, "rowTypeSymbol is null");
+        checkArgument(step.isOutputPartial() || !rowTypeSymbol.isPresent(), "rowTypeSymbol is required with partial output");
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: is this check correct? It looks like it will pass if step is partial AND rowTypeSymbol is missing.

File: presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java [hunk 5]
```diff
@@ -176,7 +177,16 @@ public Page getOutput()
         }
 
         state = State.FINISHED;
-        return pageBuilder.build();
+
+        Page page = pageBuilder.build();
+
+        // prepend the row type block for partial outputs
+        if (partial) {
+            BlockBuilder blockBuilder = TINYINT.createBlockBuilder(null, 1);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java]: A future commit introduces a `RowType` enum. Is there any reason that change is not part of this commit?

File: presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java [hunk 6]
```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
...
+        @Override
+        public Page getFinalOutput()
+        {
+            List<Block> finalResults = new ArrayList<>();
+            for (InternalSingleAccumulator accumulator : accumulators) {
+                BlockBuilder blockBuilder = accumulator.getFinalType().createBlockBuilder(null, 1);
+                accumulator.evaluateFinal(blockBuilder);
+                finalResults.add(blockBuilder.build());
+            }
+            return new Page(1, finalResults.stream().toArray(Block[]::new));
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java]: `SelectedPositions#selectPositions` creates an array of `Block` from the beginning. This code and other places in this commit create a `List<Block>` first, fill it up, then convert it to an array. Why are these differences?

File: presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java [hunk 7]
```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
...
+                    .filter(InternalAccumulatorFactory::isDistinct)
+                    .count();
+            this.intermediateIsDistinct = accumulatorFactories.stream()
+                    .map(InternalAccumulatorFactory::isDistinct)
+                    .collect(toImmutableList());
+            this.intermediateTypes = accumulatorFactories.stream()
+                    .map(InternalAccumulatorFactory::getIntermediateType)
+                    .collect(toImmutableList());
+
+            this.inputs = ImmutableList.copyOf(requireNonNull(inputs, "inputs is null"));
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java]: this check is a duplicate of the one on L481

File: presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java [hunk 8]
```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
...
         {
-            if (step.isInputRaw()) {
-                aggregation.addInput(page);
+            checkElementIndex(accumulatorIndex, intermediateTypes.size());
+            return 1 + accumulatorIndex;
+        }
+
+        private Optional<Page> buildPartialDistinctRows(Page page, List<Optional<Block>> partialDistinctMasks)
+        {
+            checkArgument(partialDistinctMasks.size() == intermediateTypes.size());
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java]: The callers of this method seem to be adding items to `partialDistinctMasks` only for distinct aggregations, hence, it can be that `partialDistinctMasks.size() < intermediateTypes.size()`, but `partialDistinctMasks.size() == intermediateIsDistinctCount` should hold.

File: presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java [hunk 9]
```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
...
+            verify(partialDistinctMasks.stream().filter(Optional::isPresent).allMatch(mask -> mask.get().getPositionCount() == page.getPositionCount()));
+
+            // check if any positions are distinct, this shouldn't happen since functions shouldn't return an empty mask
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java]: my reading of this comment is that it shouldn't happen that any position is distinct, which is probably not what it is intended

File: presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java [hunk 10]
```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
...
+            // add row type
+            blocks.add(RowType.DISTINCT_PARTIAL.createRowTypeBlock(page.getPositionCount()));
+
+            // add intermediates
+            for (int intermediateIndex = 0; intermediateIndex < partialDistinctMasks.size(); intermediateIndex++) {
+                Optional<Block> partialDistinctMask = partialDistinctMasks.get(intermediateIndex);
+                if (partialDistinctMask.isPresent()) {
+                    blocks.add(partialDistinctMask.get());
+                }
+                else if (intermediateIsDistinct.get(intermediateIndex)) {
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java]: I might be missing something, but it seems to me that `intermediateIndex` may not be aligned with `intermediateIsDistinct`

File: presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java [hunk 11]
```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
...
+                partialDistinctMasks.add(accumulator.addInput(inputPage));
+                memorySize += accumulator.getEstimatedSize();
+            }
+            systemMemoryContext.setBytes(memorySize);
 
-        Page page = pageBuilder.build();
+            if (!intermediateLayout.hasDistinct()) {
+                return Optional.empty();
+            }
+            return intermediateLayout.buildPartialDistinctRows(inputPage, partialDistinctMasks);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java]: It looks like when there are lots of duplicate values, we'll be generating very small pages. E.g. it appears that we generate one output page for every input page. Is this the intended behavior?

File: presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java [hunk 12]
```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
...
+            return SelectedPositions.positionsList(distinctPositions, 0, distinctCount);
         }
 
-        public long getEstimatedSize()
+        private Page buildIntermediateOutput(List<Block> intermediateResults)
         {
-            return aggregation.getEstimatedSize();
+            List<Block> blocks = new ArrayList<>(getOutputChannelCount());
+            blocks.add(RowType.INTERMEDIATE.createRowTypeBlock(1));
+            blocks.addAll(intermediateResults);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java]: Am I understanding correctly that pages with RowType.DISTINCT_PARTIAL have different scheme then pages with RowType.INTERMEDIATE?

RowType.DISTINCT_PARTIAL pages have rowType, followed by one boolean block per distinct aggregation, followed by  one block per input symbol.

For `select count(distinct a), count(distinct b), sum(c) from t`, RowType.DISTINCT_PARTIAL page has 1 + 2 + 3 = 6 blocks: `rowType`, `is_distinct_a:boolean`, `is_distinct_b(boolean)`, `a`, `b`, `c`.

RowType.INTERMEDIATE pages have rowType, followed by one block with partial result per aggregation, followed by some null blocks.

For above query, we have `rowType`, `partial-count(distinct a)`, `partial-count(distinct b)`, `partial-sum(c)`, `null`, `null`, `null`.

**Comment by sopel39** [comment]
I'm interested in that so I will try to review it. I'm not blocking it though
